catalog/update.go

```diff
@@ -0,0 +1,40 @@
+package catalog
+
+import (
+	"context"
+
+	"github.com/databrickslabs/terraform-provider-databricks/common"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func updateFunctionFactory(securable string, updatable []string) func(context.Context, *schema.ResourceData, *common.DatabricksClient) error {
+	return func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+		patch := map[string]interface{}{}
+		for _, field := range updatable {
+			_, new := d.GetChange(field)
+			if !d.HasChange(field) {
+				continue
+			}
+			patch[field] = new
+		}
+		if len(patch) == 0 {
+			return nil
+		}
+		switch securable {
+		case "metastore":
+			return NewMetastoresAPI(ctx, c).updateMetastore(d.Id(), patch)
+		case "catalog":
+			return NewCatalogsAPI(ctx, c).updateCatalog(d.Id(), patch)
+		case "schema":
+			return NewSchemasAPI(ctx, c).updateSchema(d.Id(), patch)
+		case "table":
+			return NewTablesAPI(ctx, c).updateTable(d.Id(), patch)
+		case "external-location":
+			return NewExternalLocationsAPI(ctx, c).update(d.Id(), patch)
+		case "storage-credential":
+			return NewStorageCredentialsAPI(ctx, c).update(d.Id(), patch)
```

Comment:
i don't like this hardcoding and tight coupling. at the same time this makes it more readable.

there are also two other options: 

very explicit:
```go
return updateFactory(func (ctx context.Context) func(string, map[string]interface{}) error {
  return NewCatalogsAPI(ctx, c).updateCatalog
}, resourceData, client, "owner", "name", "comment", "properties")
```

layer leak:
```go
update := updateFactory("/unity-catalog/catalogs", "owner", "name", "comment", "properties")
```

which would entail implementation like
```go
func updateFunctionFactory(pathPrefix string, updatable string...) func(context.Context, *schema.ResourceData, *common.DatabricksClient) error {
  return func(... {
    ...
    return c.Patch(ctx, path.Join(pathPrefix, d.Id()), patch)
  }
}
```

with this layer leak we technically don't need update methods for all securables, right?

Reply:
moved to layer leak - looking relatively clean